strawberry/custom_scalar.py

```diff
@@ -31,42 +33,80 @@ def identity(x):
     return x
 
 
+# TODO: Maybe rename this to StrawberryScalar?
 @dataclass
 class ScalarDefinition(StrawberryType):
-    name: str
-    description: Optional[str]
-    specified_by_url: Optional[str]
-    serialize: Optional[Callable]
-    parse_value: Optional[Callable]
-    parse_literal: Optional[Callable]
+    origin: Optional[type] = None
+    name: str = None
+    description: Optional[str] = None
+    specified_by_url: Optional[str] = None
+    serialize: Optional[Callable] = None
+    parse_value: Optional[Callable] = None
+    parse_literal: Optional[Callable] = None
     directives: Iterable[object] = ()
 
     # Optionally store the GraphQLScalarType instance so that we don't get
     # duplicates
     implementation: Optional[GraphQLScalarType] = None
 
-    def copy_with(
-        self, type_var_map: Mapping[TypeVar, Union[StrawberryType, type]]
-    ) -> Union[StrawberryType, type]:
-        return super().copy_with(type_var_map)
+    def to_graphql_core(self):
+        if self.implementation:
+            return self.implementation
+        else:
+            from strawberry.schema.schema_converter import GraphQLCoreConverter
+
+            res = GraphQLScalarType(
+                name=self.name,
+                description=self.description,
+                specified_by_url=self.specified_by_url,
+                serialize=self.serialize,
+                parse_value=self.parse_value,
+                parse_literal=self.parse_literal,
+                extensions={GraphQLCoreConverter.DEFINITION_BACKREF: self},
+            )
+            # cache
+            self.implementation = res
+            self.implementation = cast(GraphQLScalarType, res)
+            return self.implementation
+
+    def __call__(self, class_or_value: Any):
+        if (
+            inspect.isclass(class_or_value)
+            or type(class_or_value) is NewType
+            and not self.origin
+        ):
+            self.origin = class_or_value
+            if not self.name:
+                return dataclasses.replace(
+                    self, name=to_camel_case(class_or_value.__name__)
+                )
+            return self
+        else:
+            # if someone tries to `initialize` the scalar just return what he gave
+            # parse_value will be called else where.
+            return class_or_value
 
     @property
     def is_generic(self) -> bool:
         return False
 
+    def validate(self, value):
+        with contextlib.suppress(Exception):
+            parsed = self.parse_value(value)
+            if self.serialize(parsed) == value:
+                return True
+        with contextlib.suppress(Exception):
+            serialized = self.serialize(value)
+            if self.parse_value(serialized) == value:
+                return True
+        return False
 
-class ScalarWrapper:
```

Comment:
why did you remove this? :)

Reply:
check my comment: https://github.com/strawberry-graphql/strawberry/pull/2117#issuecomment-1231474766

> I removed ScallarWrapper as it made issued for me with the new validate method of StrawberryType, every scalar now is just a ScalarDefinition, I would consider renaming this to StrawberryScalar.

Reply:
is it needed for this PR?

Reply:
Yes. it is how generic unions would be resolved here.

https://github.com/strawberry-graphql/strawberry/pull/2117/files#diff-111023ca6e4b982dfea8bea22a5eafe990b8639010fc30dbd7e515a96fc16fb2R120